vizro-core/tests/unit/vizro/actions/_action_loop/test_get_action_loop_components.py

```diff
@@ -0,0 +1,205 @@
+        id="action_trigger_actions_id",
+        data=list(actions_ids),
+    )
+
+
+@pytest.fixture
+def trigger_to_actions_chain_mapper_component(request):
+    components = request.param
+    actions_chain_ids = [model_manager[component].actions[0].id for component in components]
+    return dcc.Store(
+        id="trigger_to_actions_chain_mapper",
+        data={
+            actions_chain_id: [action.id for action in model_manager[actions_chain_id].actions]
+            for actions_chain_id in actions_chain_ids
+        },
+    )
+
+
+@pytest.fixture
+def managers_one_page_two_components_two_controls():
+    """Instantiates managers with one page that contains two controls and two components."""
+    page = vm.Page(
+        id="test_page",
+        title="First page",
+        components=[
+            vm.Graph(
+                id="scatter_chart",
+                figure=px.scatter(px.data.gapminder(), x="lifeExp", y="gdpPercap"),
+            ),
+            vm.Button(
+                id="export_data_button",
+                actions=[vm.Action(id="export_data_action", function=export_data())],
+            ),
+        ],
+        controls=[
+            vm.Filter(id="filter_continent", column="continent", selector=vm.Dropdown(id="filter_continent_selector")),
+            vm.Parameter(
+                id="parameter_x",
+                targets=["scatter_chart.x"],
+                selector=vm.Dropdown(
+                    id="parameter_x_selector",
+                    options=["lifeExp", "gdpPercap", "pop"],
+                ),
+            ),
+        ],
+    )
+    # TODO: Call the Dashboard._pre_build() method once the pages registration is moved into this method.
+    yield Vizro().build(vm.Dashboard(pages=[page]))
+    del dash.page_registry["test_page"]
+
+
+@pytest.fixture
+def managers_one_page_no_actions():
+    """Instantiates managers with one "empty" page."""
+    page = vm.Page(
+        id="test_page_no_actions",
+        title="Second page",
+        components=[vm.Card(text="")],
+    )
+    # TODO: Call the Dashboard._pre_build() method once the pages registration is moved into this method.
+    yield Vizro().build(vm.Dashboard(pages=[page]))
+    del dash.page_registry["test_page_no_actions"]
+
+
+class TestGetActionLoopComponents:
+    """Tests getting required components for the action loop."""
+
+    @pytest.mark.usefixtures("managers_one_page_no_actions")
+    def test_no_components(self):
+        result = _get_action_loop_components()
+        assert result == []
+
+    @pytest.mark.usefixtures("managers_one_page_two_components_two_controls")
+    def test_fundamental_components(self, fundamental_components):
+        result = _get_action_loop_components()
+        for component in fundamental_components:
+            assert repr(component) in map(repr, result), f"Can't find component: {repr(component)} in: {repr(result)}"
+
+    @pytest.mark.usefixtures("managers_one_page_two_components_two_controls")
```

Comment:
Lots of these tests look very similar but with different fixtures. Would there be some better way of grouping them together into one test through parameterisation? Don't do it if it makes it much more complicated and difficult to understand though - repetition like you have it here would be better than that.

Reply:
I removed tests that individually test parts of the `_get_action_loop_components` result. Now they are all combined into one `test_all_action_loop_components` test.

P.S. I get rid of `repr` here as well.